presto-main/src/test/java/com/facebook/presto/execution/buffer/TestArbitraryOutputBuffer.java

```diff
@@ -0,0 +1,957 @@
+        assertFutureIsDone(secondEnqueuePage);
+    }
+
+    @Test
+    public void testBufferCompletion()
+            throws Exception
+    {
+        ArbitraryOutputBuffer buffer = createArbitraryBuffer(createInitialEmptyOutputBuffers(ARBITRARY), sizeOfPages(5));
+        buffer.setOutputBuffers(createInitialEmptyOutputBuffers(ARBITRARY)
+                .withBuffer(FIRST, 0)
+                .withNoMoreBufferIds());
+
+        assertFalse(buffer.isFinished());
+
+        // fill the buffer
+        List<Page> pages = new ArrayList<>();
+        for (int i = 0; i < 5; i++) {
+            Page page = createPage(i);
+            addPage(buffer, page);
+            pages.add(page);
+        }
+
+        buffer.setNoMorePages();
+
+        // get and acknowledge 5 pages
+        assertBufferResultEquals(TYPES, getBufferResult(buffer, FIRST, 0, sizeOfPages(5), MAX_WAIT), bufferResult(0, pages));
+
+        // buffer is not finished
+        assertFalse(buffer.isFinished());
+
+        // there are no more pages and no more buffers, but buffer is not finished because it didn't receive an acknowledgement yet
+        assertFalse(buffer.isFinished());
+
+        // ask the buffer to finish
+        buffer.abort(FIRST);
+
+        // verify that the buffer is finished
+        assertTrue(buffer.isFinished());
+    }
+
+    public static BufferResult getBufferResult(OutputBuffer buffer, OutputBufferId bufferId, long sequenceId, DataSize maxSize, Duration maxWait)
+    {
+        CompletableFuture<BufferResult> future = buffer.get(bufferId, sequenceId, maxSize);
+        return getFuture(future, maxWait);
+    }
+
+    public static BufferResult getFuture(CompletableFuture<BufferResult> future, Duration maxWait)
+    {
+        return tryGetFutureValue(future, (int) maxWait.toMillis(), MILLISECONDS).get();
+    }
+
+    private static ListenableFuture<?> enqueuePage(OutputBuffer buffer, Page page)
+    {
+        ListenableFuture<?> future = buffer.enqueue(page);
+        assertFalse(future.isDone());
+        return future;
+    }
+
+    private static void addPage(OutputBuffer buffer, Page page)
+    {
+        assertTrue(buffer.enqueue(page).isDone(), "Expected add page to not block");
+    }
+
+    private static void assertQueueState(
+            OutputBuffer buffer,
+            int unassignedPages,
+            OutputBufferId bufferId,
+            int bufferedPages,
+            int pagesSent)
+    {
+        OutputBufferInfo outputBufferInfo = buffer.getInfo();
+
+        long assignedPages = outputBufferInfo.getBuffers().stream().mapToInt(BufferInfo::getBufferedPages).sum();
+
+        assertEquals(
+                outputBufferInfo.getTotalBufferedPages() - assignedPages,
+                unassignedPages,
+                "unassignedPages");
+
+        BufferInfo bufferInfo = outputBufferInfo.getBuffers().stream()
+                .filter(info -> info.getBufferId().equals(bufferId))
+                .findAny()
+                .orElse(null);
```

Comment:
Isn't this the same as the helper method: getBufferInfo? Either remove that method or use it here.